File: phonebridge/src/shell/java/amirz/shade/phonebridge/shell/ProjectionShellWorkarounds.java
package amirz.shade.phonebridge.shell;

import android.app.Application;
import android.content.Context;
import android.content.ContextWrapper;
import android.content.pm.ApplicationInfo;
import android.os.Build;
import android.os.Looper;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;

final class ProjectionShellWorkarounds {
    private static final int ANDROID_12 = 31;
    private static final Class<?> ACTIVITY_THREAD_CLASS;
    private static final Object ACTIVITY_THREAD;

    static {
        prepareMainLooper();
        try {
            ACTIVITY_THREAD_CLASS = Class.forName("android.app.ActivityThread");
            Constructor<?> constructor = ACTIVITY_THREAD_CLASS.getDeclaredConstructor();
            constructor.setAccessible(true);
            ACTIVITY_THREAD = constructor.newInstance();

            Field currentThreadField = ACTIVITY_THREAD_CLASS.getDeclaredField("sCurrentActivityThread");
            currentThreadField.setAccessible(true);
            currentThreadField.set(null, ACTIVITY_THREAD);
        } catch (Exception e) {
            throw new AssertionError(e);
        }
    }

    private ProjectionShellWorkarounds() {
    }

    static void apply() {
        fillConfigurationController();
        fillAppInfo();
        fillAppContext();
    }

    static Context getSystemContext() {
        try {
            Method method = ACTIVITY_THREAD_CLASS.getDeclaredMethod("getSystemContext");
            return (Context) method.invoke(ACTIVITY_THREAD);
        } catch (Exception e) {
            throw new IllegalStateException("Unable to obtain system context", e);
        }
    }

    private static void prepareMainLooper() {
        if (Looper.myLooper() == null) {
            Looper.prepareMainLooper();
        }
    }

    private static void fillAppInfo() {
        try {
            Class<?> appBindDataClass = Class.forName("android.app.ActivityThread$AppBindData");
            Constructor<?> constructor = appBindDataClass.getDeclaredConstructor();
            constructor.setAccessible(true);
            Object appBindData = constructor.newInstance();

            ApplicationInfo applicationInfo = new ApplicationInfo();
            applicationInfo.packageName = ProjectionShellContext.SHELL_PACKAGE_NAME;

            Field appInfoField = appBindDataClass.getDeclaredField("appInfo");
            appInfoField.setAccessible(true);
            appInfoField.set(appBindData, applicationInfo);

            Field boundApplicationField = ACTIVITY_THREAD_CLASS.getDeclaredField("mBoundApplication");
            boundApplicationField.setAccessible(true);
            boundApplicationField.set(ACTIVITY_THREAD, appBindData);
        } catch (Throwable ignored) {
        }
    }

    private static void fillAppContext() {
        try {
            Application application = new Application();
            Field baseField = ContextWrapper.class.getDeclaredField("mBase");
            baseField.setAccessible(true);
            baseField.set(application, ProjectionShellContext.get());

            Field initialApplicationField = ACTIVITY_THREAD_CLASS.getDeclaredField("mInitialApplication");
            initialApplicationField.setAccessible(true);
            initialApplicationField.set(ACTIVITY_THREAD, application);
        } catch (Throwable ignored) {
        }
    }

    private static void fillConfigurationController() {
        if (Build.VERSION.SDK_INT < ANDROID_12) {
            return;
        }
        try {
            Class<?> controllerClass = Class.forName("android.app.ConfigurationController");
            Class<?> internalClass = Class.forName("android.app.ActivityThreadInternal");
            Constructor<?> constructor = controllerClass.getDeclaredConstructor(internalClass);
            constructor.setAccessible(true);
            Object controller = constructor.newInstance(ACTIVITY_THREAD);

            Field field = ACTIVITY_THREAD_CLASS.getDeclaredField("mConfigurationController");
            field.setAccessible(true);
            field.set(ACTIVITY_THREAD, controller);
        } catch (Throwable ignored) {
        }
    }
}

File: phonebridge/src/main/java/amirz/shade/phonebridge/shell/ProjectionShellWorkarounds.kt
package amirz.shade.phonebridge.shell

import android.annotation.SuppressLint
import android.app.Application
import android.content.Context
import android.content.ContextWrapper
import android.content.pm.ApplicationInfo
import android.os.Build
import android.os.Looper

@SuppressLint(
    "PrivateApi",
    "BlockedPrivateApi",
    "SoonBlockedPrivateApi",
    "DiscouragedPrivateApi"
)
internal object ProjectionShellWorkarounds {
    private val activityThreadClass: Class<*> by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
        prepareMainLooper()
        Class.forName("android.app.ActivityThread")
    }

    private val activityThread: Any by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
        val constructor = activityThreadClass.getDeclaredConstructor()
        constructor.isAccessible = true
        constructor.newInstance().also { instance ->
            val currentActivityThreadField =
                activityThreadClass.getDeclaredField("sCurrentActivityThread")
            currentActivityThreadField.isAccessible = true
            currentActivityThreadField.set(null, instance)
        }
    }

    fun apply() {
        fillConfigurationController()
        fillAppInfo()
        fillAppContext()
    }

    fun systemContext(): Context {
        return activityThreadClass.getDeclaredMethod("getSystemContext")
            .invoke(activityThread) as? Context
            ?: throw IllegalStateException("Unable to obtain system context")
    }

    @Suppress("DEPRECATION")
    private fun prepareMainLooper() {
        if (Looper.myLooper() == null) {
            Looper.prepareMainLooper()
        }
    }

    private fun fillAppInfo() {
        runCatching {
            val appBindDataClass = Class.forName("android.app.ActivityThread\$AppBindData")
            val appBindDataConstructor = appBindDataClass.getDeclaredConstructor()
            appBindDataConstructor.isAccessible = true
            val appBindData = appBindDataConstructor.newInstance()

            val applicationInfo = ApplicationInfo().apply {
                packageName = ProjectionShellContext.SHELL_PACKAGE_NAME
            }
            val appInfoField = appBindDataClass.getDeclaredField("appInfo")
            appInfoField.isAccessible = true
            appInfoField.set(appBindData, applicationInfo)

            val boundApplicationField = activityThreadClass.getDeclaredField("mBoundApplication")
            boundApplicationField.isAccessible = true
            boundApplicationField.set(activityThread, appBindData)
        }
    }

    private fun fillAppContext() {
        runCatching {
            val app = Application()
            val baseField = ContextWrapper::class.java.getDeclaredField("mBase")
            baseField.isAccessible = true
            baseField.set(app, ProjectionShellContext.instance)

            val initialApplicationField = activityThreadClass.getDeclaredField("mInitialApplication")
            initialApplicationField.isAccessible = true
            initialApplicationField.set(activityThread, app)
        }
    }

    private fun fillConfigurationController() {
        if (Build.VERSION.SDK_INT < ANDROID_12) {
            return
        }
        runCatching {
            val controllerClass = Class.forName("android.app.ConfigurationController")
            val internalClass = Class.forName("android.app.ActivityThreadInternal")
            val constructor = controllerClass.getDeclaredConstructor(internalClass)
            constructor.isAccessible = true
            val controller = constructor.newInstance(activityThread)
            val field = activityThreadClass.getDeclaredField("mConfigurationController")
            field.isAccessible = true
            field.set(activityThread, controller)
        }
    }

    private const val ANDROID_12 = 31
}
